ledger/participant-state/kvutils/src/main/scala/com/daml/ledger/participant/state/kvutils/export/SerializationBasedLedgerDataExporter.scala

```diff
@@ -0,0 +1,70 @@
+// Copyright (c) 2020 Digital Asset (Switzerland) GmbH and/or its affiliates. All rights reserved.
+// SPDX-License-Identifier: Apache-2.0
+
+package com.daml.ledger.participant.state.kvutils.export
+
+import java.io.DataOutputStream
+import java.time.Instant
+import java.util.concurrent.locks.StampedLock
+
+import com.daml.ledger.validator.LedgerStateOperations.{Key, Value}
+
+/**
+  * Enables exporting ledger data to an output stream.
+  * This class is thread-safe.
+  */
+final class SerializationBasedLedgerDataExporter(output: DataOutputStream)
+    extends LedgerDataExporter {
+
+  private val outputLock = new StampedLock
+
+  override def addSubmission(submissionInfo: SubmissionInfo): SubmissionAggregator = {
+    new InMemorySubmissionAggregator(submissionInfo, SerializationBasedLedgerDataWriter)
+  }
+
+  object SerializationBasedLedgerDataWriter extends LedgerDataWriter {
+    override def write(submissionInfo: SubmissionInfo, writeSet: Seq[(Key, Value)]): Unit = {
+      val stamp = outputLock.writeLock()
+      try {
+        serializeEntry(submissionInfo, writeSet)
+        output.flush()
+      } finally {
+        outputLock.unlock(stamp)
+      }
+    }
+
+    private def serializeEntry(
+        submissionInfo: SubmissionInfo,
+        writeSet: Seq[(Key, Value)],
+    ): Unit = {
+      serializeSubmissionInfo(submissionInfo)
+      serializeWriteSet(writeSet)
+    }
+
+    private def serializeSubmissionInfo(submissionInfo: SubmissionInfo): Unit = {
+      output.writeUTF(submissionInfo.correlationId)
+      writeBytes(submissionInfo.submissionEnvelope)
+      writeInstant(submissionInfo.recordTimeInstant)
+      output.writeUTF(submissionInfo.participantId)
+    }
+
+    private def serializeWriteSet(writeSet: WriteSet): Unit = {
+      output.writeInt(writeSet.size)
+      for ((key, value) <- writeSet.sortBy(_._1.asReadOnlyByteBuffer())) {
```

Comment:
Please don't sort the write-set here. I.e., this way we are hiding the fact if the producer of the write-set is non-deterministic.

Reply:
Why does it matter if it's sorted here or just before writing to the file?

This code was simply moved from https://github.com/digital-asset/daml/pull/7258/files#diff-11c44df5a2377a0c67fbbbb179af44ebL88.

Reply:
Sorry, wrong location. It comes from https://github.com/digital-asset/daml/pull/7258/files#diff-4db1eab737dbe983bba6541fe2b1ec34L33.

Reply:
It matters because then we won't know if we generated the baseline in the wrong order or not.